ldc: 1.24.0 -> 1.25.1 update (fixes broken) ldc and enable dub to build with alternative D compilers #115028

ralph-amissah opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

ldc: 1.24.0 -> 1.25.1 (fixes broken) ldc due to new llvm

dub: enable to build with alternative D compilers

fixes issue: https://github.com/NixOS/nixpkgs/issues/114875 (see discussion)
replaces pull: https://github.com/NixOS/nixpkgs/pull/115028 (as discussed)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

ralph-amissah commented:
re: dub, not sure about the changes to dub default nix between the two patches posted; implemented request without double checking and only tested  now (sorry) but: removing the -f flag (using only -r instead of -rf) as requested in #115028  SuperSandro2000 results in a build failure for: nix-shell -p dub --pure
(it builds when the -f flag is retained) should/may I revert? (i.e. reintroduce -rf)
see: https://github.com/NixOS/nixpkgs/pull/115028

ralph-amissah commented:
ok, read the error message understood the reason for SuperSandro2000 request removing one non-existent file fixes the issue

ralph-amissah commented:
ldc: ok
dub: ok sorted (SuperSandro2000 request complied with (s/-rf/-r/))

i077 commented:
You seem to be force-pushing your branch incorrectly, because other people's commits should not be in this PR. Either drop those commits or rebase your two commits on top of master.

ralph-amissah commented:
heh, yes i did a few things without being aware of what the consequences would be, pulling to master trying to rebase to my D-lang branch, and aborting that ... i should just have force pushed by branch

ralph-amissah committed:
ldc: 1.24.0 -> 1.25.1

- update to current ldc

- fixes issue #114875 broken ldc (D) compiler due to update to
  LLVM 11.1.0 (further discussion in closed pull request #115028)

ralph-amissah committed:
dub: enable to build with alternative D compilers

- enable dub (D build tool) optionally to build
  with one of the alternative D compilers, as the
  current default dmd is not what every D project
  would choose to use, and it should not be
  necessary to pull in dmd just to build dub
  (assuming dub is used)

- dmd (and in consequence the provided dub
  default.nix) have been broken in nixpkgs
  unstable for months now; meaning: the only way
  to build a D project that relies on dub is to
  use an alternative compiler to dmd such as ldc2
  (ldmd2) as this patch makes possible

- related discussions in issues #114875 and closed
  pull request #115028

SuperSandro2000 commented on pkgs/development/tools/build-managers/dub/default.nix:
```diff
@@ -1,4 +1,6 @@
-{ lib, stdenv, fetchFromGitHub, curl, dmd, libevent, rsync }:
+{ lib, stdenv, fetchFromGitHub, curl, libevent, rsync, ldc ? null, dcompiler ? ldc }:
```
```suggestion
{ lib, stdenv, fetchFromGitHub, curl, libevent, rsync, ldc, dcompiler ? ldc }:
```

ralph-amissah replied:
SuperSandro2000  will do, but you are welcome to take it, fix it and merge :-p ;-) :-)
thanks for your effort and patience all, i077 of course that means you in particular too  (hope it is not presumptuous to assume this is done eventually :)

i077 replied:
Thanks for taking the initiative on this!

It's generally not a good idea to leave it to the reviewer to make the change, they typically do a lot of these :P
Could you just make the change and force-push? Thanks!

ralph-amissah replied:
thought i already had ... think i have already ...
(point taken though) (thanks for the mentoring)

i077 commented:
Result of `nixpkgs-review pr 115133` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>Literate</li>
    <li>lumpy</li>
    <li>sambamba</li>
  </ul>
</details>
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>cheesecutter</li>
    <li>dub</li>
    <li>ldc</li>
    <li>onedrive</li>
    <li>rund</li>
  </ul>
</details>

i077 commented:
Failing packages are because dmd is failling to build (which we already know). I guess you can either remove/replace the dmd dependency, or just leave them as is (not really sure what the proper thing is to do here, heh).

---

Removing dmd from the list of buildInputs for sambamba seems to fix sambamba and lumpy:

```diff
diff --git a/pkgs/applications/science/biology/sambamba/default.nix b/pkgs/applications/science/biology/sambamba/default.nix
index 8802736e..96993da7 100644
--- a/pkgs/applications/science/biology/sambamba/default.nix
+++ b/pkgs/applications/science/biology/sambamba/default.nix
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchFromGitHub, python3, which, dmd, ldc, zlib }:
+{ lib, stdenv, fetchFromGitHub, python3, which, ldc, zlib }:
 
 stdenv.mkDerivation rec {
   pname = "sambamba";
@@ -12,7 +12,7 @@ stdenv.mkDerivation rec {
     fetchSubmodules = true;
   };
 
-  nativeBuildInputs = [ which python3 dmd ldc ];
+  nativeBuildInputs = [ which python3 ldc ];
   buildInputs = [ zlib ];
 
   # Upstream's install target is broken; copy manually

```

---

Similarly with Literate,

```diff
diff --git a/pkgs/development/tools/literate-programming/Literate/default.nix b/pkgs/development/tools/literate-programming/Literate/default.nix
index 87213b91..6174fc0c 100644
--- a/pkgs/development/tools/literate-programming/Literate/default.nix
+++ b/pkgs/development/tools/literate-programming/Literate/default.nix
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchgit, dmd, dub }:
+{ lib, stdenv, fetchgit, ldc, dub }:
 
 stdenv.mkDerivation {
   pname = "Literate";
@@ -10,7 +10,7 @@ stdenv.mkDerivation {
     sha256 = "09h1as01z0fw0bj0kf1g9nlhvinya7sqq2x8qb6zmhvqqm6v4n49";
   };
 
-  buildInputs = [ dmd dub ];
+  buildInputs = [ ldc dub ];
 
   installPhase = "install -D bin/lit $out/bin/lit";
```

ralph-amissah commented:
I don't know either. Clearly dmd's build needs fixing. 
We need the likes of @lionello and @ThomasMader
Thomas Mader has been a proponent of using Nix with D from before Nix* really had entered my consciousness  and Lio knows lots about D on Nix, (he must to do what he has done). There is also @Marler of rund fame. Ideally these three and others interested in D should agree the direction to take. I am sure there are others I do not know, (my interest has been narrow, I love my NixOS environment experience, and have my project that I wish to work on (without sacrificing being able to use NixOS)). I do not know Nix in a way that would allow me to continue the work on building D compilers using Nix in the way they have or even contribute particularly meaningfully and hope they continue to use D on Nix and do the heavy lifting ... it seems more manpower is needed just to keep things up to date based on the work already done. (that said, I am chuffed that the little dub default.nix fix might help any other D project on Nix, wonderful, I should have thought to complain sooner!).

There soon will be 3 D compilers with different strengths ... ideally any one could be chosen, and such flexibility should be preserved, but perhaps there should be a default. Though not much of an issue now I believe D libraries may need to be built by the same compiler as the calling program for it to build, which has implications for the storing of binaries for reuse. Which of the 3 should be the Nix default, there are arguments for each:
- dmd is reference and very fast in its compilation, good also for scripting; 
- ldc produces cleaner faster binaries (than dmd) and has been this way for a while, it targets more architectures (llvm)
- gdc/gcc will produce cleaner faster binaries (than dmd), and targets more architectures (gcc) ... it has been behind current dmd
**My vote at least for the time being is for ldc as current default, to be reviewed in light of gcc once that is in full flight.**
While on the subject Lio has dub2nix somewhere on github, hmm, ok it is here: https://github.com/lionello/dub2nix but it is not in nixpkgs and it should probably be at very least to make it more easily discoverable, I happened to be told of it by adisbladis on irc.freenode #nixos (or would not have thought to have looked for that specific name) ... dub2nix has also recently been discussed on the general D forum, so how to copy Lio and request that he adds it.

To the specific question at hand, matching the pattern I have a suggestion to use the dcompiler sequence currently in dub that I have pushed to branch D-lang-other at https://github.com/ralph-amissah/nixpkgs essentially:

for Literate:
```
- { lib, stdenv, fetchgit, dmd, dub }:
+ { lib, stdenv, fetchgit, ldc, dcompiler ? ldc, dub }:
```
and
```
- buildInputs = [ dmd dub ];
+ buildInputs = [ dcompiler dub ];
```

and for sambamba:
```
- { lib, stdenv, fetchFromGitHub, python3, which, dmd, ldc, zlib }:
+ { lib, stdenv, fetchFromGitHub, python3, which, ldc, dcompiler ? ldc, zlib }:
```
and
```
- nativeBuildInputs = [ which python3 dmd ldc ];
+ nativeBuildInputs = [ which python3 dcompiler ];
```

but I do not know whether this is the right or best way, or even as yet fully comprehend the implications of this choice for the future. Those who do, please chime in!

DMD in Nixpkgs needs to be updated, it is broken, that needs to be flagged ... an issue report should be filed

i077 commented:
So I see two separate issues here:

- DMD is currently (and has been for a while) broken on nixpkgs.
- It may be desirable to pick, for each D package, which D compiler to use.

I think these can be tracked in a separate issue outside of this PR though, since the current PR doesn't actually _create_ any new breakages in any packages. So this PR I think can (and should) be merged in.

@SuperSandro2000 thoughts?

ralph-amissah commented:
@i077 well I bring up perhaps 3 issues only the first of which is immediately relevant:
- (i) this patch (a) to update (thereby fixing) ldc and (b) making it possible to have a working dub, is ready to go, (unless there are other things that need to be done for the patch); rather nicely it appears to offer a fix for a few other D related packages in nixpkgs.

and as you mention (musings brought up by me, and not of direct relevance):
- (ii) "DMD is currently (and has been for a while) broken on nixpkgs".
- (iii) "It may be desirable to pick, for each D package, which D compiler to use". _For which I wonder, whether down the line it might be possible to have a dcompiler option that can be set with the default there perhaps at ldc (or in future gdc), ...?_

So is this (i) ready to go? or is there more that needs to be seen to?

@SuperSandro2000 @i077 thanks again
